core/services/request/CurrentPage.php

```diff
@@ -0,0 +1,292 @@
+
+    /**
+     * @var string
+     */
+    private $post_name = '';
+
+    /**
+     * @var array
+     */
+    private $post_type = [];
+
+    /**
+     * @var RequestInterface $request
+     */
+    private $request;
+
+
+    /**
+     * CurrentPage constructor.
+     *
+     * @param EE_CPT_Strategy  $cpt_strategy
+     * @param RequestInterface $request
+     */
+    public function __construct(EE_CPT_Strategy $cpt_strategy, RequestInterface $request)
+    {
+        $this->cpt_strategy = $cpt_strategy;
+        $this->request      = $request;
+        $this->initialized  = is_admin();
+        // analyse the incoming WP request
+        add_action('parse_request', [$this, 'parseQueryVars'], 2, 1);
+    }
+
+
+    /**
+     * @param WP $WP
+     * @return void
+     */
+    public function parseQueryVars(WP $WP = null)
+    {
+        if ($this->initialized) {
+            return;
+        }
+        // if somebody forgot to provide us with WP, that's ok because its global
+        if (! $WP instanceof WP) {
+            global $WP;
+        }
+        $this->post_id   = $this->getPostId($WP);
+        $this->post_name = $this->getPostName($WP);
+        $this->post_type = $this->getPostType($WP);
+        // true or false ? is this page being used by EE ?
+        $this->setEspressoPage();
+        remove_action('parse_request', [$this, 'parseRequest'], 2);
+        $this->initialized   = true;
+    }
+
+
+    /**
+     * Just a helper method for getting the url for the displayed page.
+     *
+     * @param WP|null $WP
+     * @return string
+     */
+    public function getPermalink(WP $WP = null)
+    {
+        $post_id = $this->post_id ?: $this->getPostId($WP);
+        if ($post_id) {
+            return get_permalink($post_id);
+        }
+        if (! $WP instanceof WP) {
+            global $WP;
+        }
+        if ($WP->request) {
+            return site_url($WP->request);
+        }
+        return esc_url_raw(site_url($_SERVER['REQUEST_URI']));
+    }
+
+
+    /**
+     * @return array
+     */
+    public function espressoPostType()
+    {
+        return array_filter(
+            $this->post_type,
+            function ($post_type) {
+                return strpos($post_type, 'espresso_') === 0;
+            }
+        );
+    }
+
+
+    /**
+     * @param WP $WP
+     * @return int
+     */
+    private function getPostId(WP $WP)
+    {
+        $post_id = null;
+        if (isset($WP->query_vars['p'])) {
+            $post_id = $WP->query_vars['p'];
+        }
+        if (! $post_id && isset($WP->query_vars['page_id'])) {
+            $post_id = $WP->query_vars['page_id'];
+        }
+        if (! $post_id && $WP->request !== null && is_numeric(basename($WP->request))) {
+            $post_id = basename($WP->request);
+        }
+        if (! $post_id && $this->request->requestParamIsSet('post_id')) {
+            $post_id = $this->request->getRequestParam('post_id');
+        }
+        return $post_id;
```

Comment:
I think this needs some comments to make it easy to understand.

Reply:
you want me to write comments for some code i wrote... like... maybe a decade ago?!?!

I'll try... 😟

Reply:
So the first glance at this is confusing to all of us 😅 So might as well leave as is 🤷‍♂️

Reply:
too late... already updated it :)